src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1205,18 +1210,22 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
                         break;
 
                     case DeclarationModifiers.Ref:
-                        // 'ref' is only a modifier if used on a ref struct
-                        // it must be either immediately before the 'struct'
-                        // keyword, or immediately before 'partial struct' if
-                        // this is a partial ref struct declaration
+                        // 'ref' is only a modifier if used on a ref struct but if it's followed by 'readonly'
+                        // we need to check the next token to disambiguate with 'ref readonly' on types
                         {
                             var next = PeekToken(1);
-                            if (next.Kind == SyntaxKind.StructKeyword ||
-                                (next.ContextualKind == SyntaxKind.PartialKeyword &&
-                                 PeekToken(2).Kind == SyntaxKind.StructKeyword))
+                            var isRefFollowedByStructOrPartial =
```

Comment:
I think we should fold this logic out into an IsRefStruct method, similar to IsPartialType. Also, I think the check for isRefFollowedByStructOrPartial is wrong: you could have partial be a type, and ref be a part of that type. For instance,

```C#
class partial {}
ref partial M() => null;
```

In general, I think we should consider being as robust as possible here by just looking ahead until we hit something that will definitely disambiguate: a type name, a struct/class/enum keyword, or a member name.

The same thing should probably be true of partial as well. That will allow us to add contextual modifiers in the future and keep having this code working without special modification.

Reply:
> think we should fold this logic out into an IsRefStruct

How about `IsPossibleTypeDeclaration` then we can produce meaningful errors when we see something like `ref class X {}`

Reply:
In that case, we should probably move the langver check to binding. Is that alright?

Reply:
in that case, we should probably move langver check to binding (or possibly somewhere else where we check for duplicated mods). is that alright?